numpy/lib/function_base.py

```diff
@@ -75,6 +76,87 @@ def iterable(y):
+    A helper function to be called from histogram to deal with estimating optimal number of bins
+
+    if estimator is a string
+        Check the corresponding estimator exists and
+        returns the optimal number of bins calculated using the estimator
+    """
+    if not isinstance(estimator, basestring):
+        raise ValueError("bins = '{0}', should be a string".format(estimator))
+    if a.size == 0:
+        return 1
+
+    def sturges(x):
+        """
+        Sturges Estimator
+        A very simplistic estimator based on the assumption of normality of the data
+        Poor performance for non-normal data, especially obvious for large X.
+        Depends only on size of the data
+        """
+        return np.ceil(np.log2(x.size)) + 1
+
+    def rice(x):
+        """
+        Rice Estimator
+        Another simple estimator, with no normality assumption.
+        It has better performance for large data, but tends to be too large
+        The number of bins is proportional to the cube root of data size (asymptotically optimal)
+        Depends only on size of the data
+        """
+        return np.ceil(2 * x.size ** (1.0 / 3))
+
+    def scott(x):
+        """
+        Scott Estimator
+        The binwidth is proportional to the standard deviation of the data and
+        inversely proportional to the cube root of data size (asymptotically optimal)
+
+        """
+        h = 3.5 * x.std() * x.size ** (-1.0 / 3)
+        if h > 0:
+            return np.ceil(x.ptp() / h)
+        return 1
+
+    def FD(x):
+        """
+        Freedman Diaconis rule using Inter Quartile Range (IQR) for binwidth
+        Considered a variation of the Scott rule with more robustness as the IQR
+        is less affected by outliers than the standard deviation.
+
+        If the IQR is 0, we return 1 for the number of bins.
+        Binwidth is inversely proportional to the cube root of data size (asymptotically optimal)
+        """
+        iqr = np.subtract(*np.percentile(x, [75, 25]))
+
+        if iqr > 0:
+            h = (2 * iqr * x.size ** (-1.0 / 3))
+            return np.ceil(x.ptp() / h)
+        else:
+            # If iqr is 0, default number of bins is 1
+            return 1
+
+    def auto(x):
+        """
+        The FD estimator is usually the most robust method, but it tends to be too small
+        for small X. The Sturges estimator is quite good for small (<1000) datasets and is
+        the default in R.
+        This method gives good off the shelf behaviour.
+        """
+        return max(FD(x), sturges(x))
+
+    optimal_numbins_methods = {'sturges': sturges, 'rice': rice, 'scott': scott,
+                                 'fd': FD, 'auto': auto}
+
+    if estimator.lower() not in optimal_numbins_methods.keys():
+        raise ValueError("{0} not a valid method for `bins`".format(estimator))
+
+    # these methods return floats, np.histogram requires an int
+    return int(optimal_numbins_methods[estimator.lower()](a))
```

Comment:
my approach here would probably be to use a try/except to avoid writing estimator.lower twice, e.g.,
```python
try:
    estimator_func = optimal_numbins_methods[estimator.lower()]
except KeyError:
    raise ValueError(...)
else:
    return int(estimator_func(a))
```

but this is really fine, too